monix-reactive/shared/src/test/scala/monix/reactive/internal/builders/UnfoldObservableSuite.scala

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2014-2019 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.reactive.internal.builders
+
+import cats.laws._
+import cats.laws.discipline._
+import monix.execution.Ack.Continue
+import monix.execution.internal.Platform
+import monix.reactive.observers.Subscriber
+import monix.reactive.{BaseTestSuite, Observable}
+import org.scalacheck.Gen
+import org.scalacheck.Prop._
+
+import scala.concurrent.duration.MILLISECONDS
+
+object UnfoldObservableSuite extends BaseTestSuite {
+
+  test("should be exception-proof") { implicit s =>
+    val dummy = new RuntimeException("dummy")
+    var received = 0
+
+    Observable.unfold(0)(i => if (i < 20) Some((i, i + 1)) else throw dummy).subscribe { _: Int =>
+      received += 1
+      Continue
+    }
+
+    assertEquals((0 until received).toList, (0 to 19).toList)
+    assertEquals(s.state.lastReportedError, dummy)
+  }
+
+  test("should execute 10 times then return None") { implicit s =>
+    var received = 0
+
+    Observable.unfold(0)(i => if (i < 10) Some((i, i + 1)) else None).subscribe { _: Int =>
+      received += 1
+      Continue
+    }
+
+    assertEquals((0 until received).toList, (0 to 9).toList)
+  }
+
+  test("should be cancelable") { implicit s =>
+    var wasCompleted = false
+    var sum = 0
+
+    val cancelable = Observable
+      .unfold(s.clockMonotonic(MILLISECONDS))(intOption)
+      .unsafeSubscribeFn(new Subscriber[Int] {
+        implicit val scheduler = s
+
+        def onNext(elem: Int) = {
+          sum += 1
+          Continue
+        }
+
+        def onComplete() = wasCompleted = true
+
+        def onError(ex: Throwable) = wasCompleted = true
+      })
+
+    cancelable.cancel()
+
+    s.tick()
+
+    assertEquals(sum, s.executionModel.recommendedBatchSize * 2)
+    assert(!wasCompleted)
+  }
+
+  test("unfold and fromStateAction results should be equal given generated inputs") { implicit s =>
+    check {
+      forAll(Gen.choose(0, Platform.recommendedBatchSize * 2), Gen.choose(0, Platform.recommendedBatchSize * 2)) {
```

Comment:
I think the test is good but I'd recommend doing just `check2` and `val n = i % (recommendedBatchSize * 2)` like in many `Iterant` tests, it's a bit less noise when reading tests

Reply:
so basically something like

```
    check2 { (s: Int, i: Int) =>
          val seed = s % (recommendedBatchSize * 2)
          val n = i % (recommendedBatchSize * 2)
```
?

Reply:
I feel like `seed` should be fine but there's no harm in doing it this way 👍